tokio/src/task/task_local.rs

```diff
@@ -0,0 +1,240 @@
+use pin_project_lite::pin_project;
+use std::cell::RefCell;
+use std::error::Error;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::{fmt, thread};
+
+/// Declare a new task local storage key of type [`tokio::task::LocalKey`].
+///
+/// # Syntax
+///
+/// The macro wraps any number of static declarations and makes them task locals.
+/// Publicity and attributes for each static are allowed.
+///
+/// # Examples
+///
+/// ```
+/// # use tokio::task_local;
+/// task_local! {
+///     pub static FOO: u32;
+///
+///     #[allow(unused)]
+///     static BAR: f32;
+/// }
+/// # fn main() {}
+/// ```
+///
+/// See [LocalKey documentation][`tokio::task::LocalKey`] for more
+/// information.
+///
+/// [`tokio::task::LocalKey`]: ../tokio/task/struct.LocalKey.html
+#[macro_export]
+macro_rules! task_local {
+     // empty (base case for the recursion)
+    () => {};
+
+    ($(#[$attr:meta])* $vis:vis static $name:ident: $t:ty; $($rest:tt)*) => {
+        $crate::__task_local_inner!($(#[$attr])* $vis $name, $t);
+        $crate::task_local!($($rest)*);
+    };
+
+    ($(#[$attr:meta])* $vis:vis static $name:ident: $t:ty) => {
+        $crate::__task_local_inner!($(#[$attr])* $vis $name, $t);
+    }
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __task_local_inner {
+    ($(#[$attr:meta])* $vis:vis $name:ident, $t:ty) => {
+        static $name: $crate::task::LocalKey<$t> = {
+            std::thread_local! {
+                static __KEY: std::cell::RefCell<Option<$t>> = std::cell::RefCell::new(None);
```

Comment:
I'm not super familiar with how thread-locals are implemented. Do they when defined use memory on every thread - even if that one doesn't make use of them? Or is it lazily allocated? If it requires memory everywhere then adding lots of task-locals will obviously make it worse. But on the other hand I don't expect people to use too many task-locals with this kind declaration mechanism.

Reply:
IIRC, thread-local pages are copy-on-write.

Reply:
https://akkadia.org/drepper/tls.pdf